### DISCUSSION TITLE
Using Test Harness with MPI on HPC

### DISCUSSION THREAD

**Discussion by Vandenbg** [idaholab/moose#25342] [Q&A General] [answered]
Hello,
I am attempting to take advantage of the Test Harness and MPI. I have a test file which contains ~1000 parameterizations for a simulation. Here is my PBS script:
```
#PBS -N gtvb_job
#PBS -l select=5:ncpus=48:mpiprocs=48
#PBS -l walltime=24:00:00
#PBS -P moose

JOB_NUM=${PBS_JOBID%%\.*}

input_file=input_test

cd /home/vandgavi/input

module load use.moose moose-dev

'/home/vandgavi/projects/moose/./run_tests' -i $input_file -C $PWD -x -t --include-input-file -j 240 -p 24
```
Before simulation start the 'Active Jobs' tab states that I have requested 5 nodes and 240 cpu. Once the simulation starts, this changes to 1 node and 48 cpu. Each simulation then runs with 'Num Processors:1, Num Threads:1', which is drastically slower than using the mpiexec command for 1 parameterization using 1 node and 24 cpu.

I have this same issue with a test file with several 'sampler, type = quadrature' calculations. These initially state the use of 1 node and 24 cpu, however, each calculation by the quadrature sample is run with 'Num Processors:1, Num Threads:1'.

Is this a bug or a feature? Is there a way to use the Test Harness with MPI?

**Comment by GiudGiud** [comment]
Hello

What you have would be fine if you had a single node with 240 CPUs essentially. 
If you do `./run_tests --help` you should see all the options for using PBS instead of relying on the python distribution of tests

Guillaume

**Reply by Vandenbg** [reply]
Hello, Guillaume. These are the options you're talking about. 
```
  Options controlling which queue manager to use
  --pbs name               Launch tests using PBS as your scheduler. You must supply a name to identify this session with
  --pbs-pre-source    Source specified file before launching tests
  --pbs-project           Identify your job(s) with this project (default: moose)
  --pbs-queue             Submit jobs to the specified queue
  --pbs-node-cpus       CPUS Per Node. The default (no setting), will always use only one node
  --pbs-cleanup name    Clean up files generated by supplied --pbs name
```
The only one I understand in this is '--pbs-node-cpus 48'. I tacked it onto the ./run_tests command within my PBS script and I am still getting the same problem.

I sort of understand your reply, could you explain the second sentence further for me?

**Reply by GiudGiud** [reply]
I dont think you should be including this in your script.
I think you just run run_tests on the command line with the pbs options to start pbs jobs

**Reply by milljm** [reply, marked as answer]
GiudGiud is correct. You use the TestHarness to launch PBS jobs, and not PBS to launch the TestHarness. Er... this isn't 100% accurate. The TestHarness _does_ invoke the TestHarness again. But in a very controlled way. The fact remains this is not supposed to be the responsibility of the user.

What you should do instead (but there may be caveats in your case):
```bash
head_node> ./run_tests --pbs my_run \
  --spec-file path/to/1000_job_test_spec_file \
  --pbs-node-cpus 48 \
  --pbs-project gtvb_job \
  --pbs-pre-source /path/to/env_profile # a bash/sh profile that performs `module load use.moose moose-dev`
```
This will create a PBS Script, and launch it via `qsub`.

The TestHarness will figure out how many nodes it needs based on the maximum discovered core count requirement among all jobs in a single test spec file (`min_parallel = `) divided by supplied `--pbs-node-cpus` argument. So in order for the TestHarness to create the same PBS submission script you supplied, somewhere among your 1000 job test spec file you will need a job asking for a minimum of 240:

```pre
[Tests]
  [test_1000]
    min_parallel = 240
  [../]
[]
```

Where the TestHarness isn't cool... is if among your 1000 jobs you have jobs only asking for 1 core, or 2, or anything less than 240, weird 'stuff' is going to happen:

The TestHarness being Python based, is only scalable to about 12 simultaneous jobs. So however you end up divvying out jobs, if the TestHarness begins to track the goings-on of more than this, you'll start to see a performance drop.

Basically, if the TestHarness ends up running `-j 240`, and in turn actually launches 240 single jobs, I cannot say how efficient that will be.

The bottom line, I would have each job in your 1000 job test spec file contain `min_parallel = 240`. This of course will mean each job (all 1000 of them) will run serially.

I hope this made some sense. I'd be happy to go into more detail!

**Reply by Vandenbg** [reply]
Thanks for the suggestion Guillaume, and the explanation Jason.

I have been tinkering with the ./run_tests command and I have a confirmation and a clarifying question. 
1. min_parallel is the number of processors used by a single simulation and to reduce errors I should set it to the same value for all simulations? If I set min_parallel = 480 and --pbs-node-cpus 48, then it will do 10 simulations at once with 48 cpu? Can I place it outside a simulation block? Like so:
```
[Tests]
  min_parallel = 240
  [test_1000]
     ...
  []
[]
```
2. For --pbs-pre-source, I should put the path to my .bash_profile which contains these lines? or would it be better to just make a short bash script with just the command?
```
# start moose every time
module load use.moose moose-dev
```

Thanks for the help!!!

**Reply by GiudGiud** [reply]
I've never put min_parallel here, always in individual tests.Does it work?

**Reply by milljm** [reply]
> min_parallel = 480 and --pbs-node-cpus 48, then it will do 10 simulations at once with 48 cpu?

This will instruct the TestHarness to request the following resources in the PBS select statement:

```pre
#PBS -l select=10:ncpus=48:mpiprocs=48
```
As for what the TestHarness will do when it encounters your tests, if you supply the following in your test spec file:
```pre
[Tests]
  min_parallel = 480
  [test_1000]
     ...
  []
[]
```
The 'second' TestHarness will receive the following `-j` argument while on a node:
```bash
./run_tests -j 480
```
Which will execute your MOOSE based application thusly:
```bash
mpiexec -n 480 moose_app-opt -i some_input_file.i
```
And because `min_parallel` is set as such for every test in the test spec file, the TestHarness will understand that it only has enough resources granted to it, to execute one job at a time.

EDIT: forgot to answer your bash/profile Q:

I would try to limit what the TestHarness's sources to an absolute minimum. But thats me! I see no harm in asking the TestHarness to source your ~/.bashrc. Complicated things _might_ get in the way I suppose (Like, if your ~/.bashrc also sources/loads Conda, etc)

**Reply by milljm** [reply]
Honestly I've never placed min_parallel in a top-level block as well. I am not sure if this works. But I also don't see why that wouldn't work.

**Reply by milljm** [reply]
To tell you the truth, the TestHarness was never intended to facilitate large jobs. Just a bunch of smaller ones. Its there to "test" the capabilities of MOOSE, not really to be your problem solver/manager. I applaud your attempt! I am curious how it works out for you.

**Reply by Vandenbg** [reply]
I appreciate the discussion and conversion to the scripts I know, this is very helpful. Placing it on top gives me a fatal error ("QSUB Group Failure"), that may be caused by something else, I'll need to trouble shoot. I have been told several times that the test harness isn't meant for this, however, it has an amazing capability that I want to take advantage of. I will place it within each block! I'll make a post later if I find anything.

Cheers!

**Reply by milljm** [reply]
I like the idea of supporting `min_parallel` at the top. I will look into adding this!

**Reply by milljm** [reply]
Created issue: https://github.com/idaholab/moose/issues/25359

**Comment by Vandenbg** [comment]
Okay, I've had a bit of time to mess with the ./run_tests PBS options. I found a few things while trouble shooting. 

This is what I submitted:
```
'/home/vandgavi/projects/bison/./run_tests' --pbs testing --spec-file /home/vandgavi/inputs/fp/di/testing_3 --pbs-node-cpus 48 --pbs-project ne_ldrd --pbs-pre-source /home/vandgavi/.bash_profile
```
`--pbs $name` this will create an output file where your run tests file is $name. 
`--pbs-project $project` required for project.
`--spec-file` this is where the disconnect begins. I tell ./run_tests to run my testing_3 spec file and it creates a `$test_name.qsub` script within the specified folder:
```
#!/bin/bash
#PBS -N params-di_1
#PBS -l select=1:ncpus=24
#PBS -l walltime=00:06:00
#PBS -P ne_ldrd

#PBS -j oe
#PBS -o /home/vandgavi/inputs/fp/di/qsub.output
#PBS -l place=free
source /home/vandgavi/.bash_profile || exit 1
JOB_NUM=${PBS_JOBID%\.*}

export MV2_ENABLE_AFFINITY=0

cd /home/vandgavi/projects/bison
/home/vandgavi/projects/bison/run_tests -j 24 --pbs-node-cpus 48 --pbs-project ne_ldrd --pbs-pre-source /home/vandgavi/.bash_profile --spec-file /home/vandgavi/inputs/fp/di/tests -o /home/vandgavi/inputs/fp/di --sep-files
```
There are a couple things to point out here and other tests I tried. 

The **walltime** requested is a sum of each max_time in the script, for example, my testing spec has 3 tests each with 120s max_time. Results in `walltime=00:06:00`.  The sum of max_time's within the spec can NOT exceed the max wall time of 168 hours, it will request over maximum. @milljm this may be something you could fix in the same issue, I am not sure. 

The next is that if the outputs of this script are present, it will automatically skip all the tests. if you wish to re-run this same test, all outputs must be deleted, i.e. the `$test_name.qsub` and `$name` files in their respective folders mentioned earlier. 

This is something I call a bug, however, it may be a feature. The created .qsub file contains the line:
`--spec-file /home/vandgavi/inputs/fp/di/tests`
This is wrong, even though my initial ./run_tests command includes the name of my spec as `testing_3` it instead ran my `tests` spec which only checks input syntax. For now you must name the spec file you want to run, `tests`. Which throws a wrench in my plans as I have roughly 70 spec files with different names I'd like to use. 

The first test in the script will be the name that shows up within the "Active Jobs" tab. This will show that only one job is running, however, the output shows that all three are running at one after another. 

Within each test block in the spec file `min_parallel = #` must be specified or it will run on one cpu. The `min_parallel=#` divided by `--pbs-node-cpus #` will show up in the `-j #` of the `$test_name.qsub` file above. It will ask for the right number in the `$test_name.qsub`, however, when it trys to cue the job with those resources, it says there are "insufficient slots". For example, it will cue 2 nodes 96 cpu, the tests with `min_parallel=96` will fail and it will say insufficient slots. Right now it seems like it can only use 1 node 48 cpu. 

At this point, I'm not sure I can use Test Harness with MPI on the HPC.

**Reply by milljm** [reply]
Can you post the contents of `/home/vandgavi/inputs/fp/di/testing_3`? I'd like to understand why the TestHarness decided to call it something different later on in the PBS script. I recall we do some separation work, but we don't create new spec files as far as I know. IIRC `-i` and `--spec-file` are supposed to be treated as precisely that: _this_ test file.

> The walltime requested is a sum of each max_time in the script, for example, my testing spec has 3 tests each with 120s max_time. Results in walltime=00:06:00. The sum of max_time's within the spec can NOT exceed the max wall time of 168 hours, it will request over maximum. @milljm this may be something you could fix in the same issue, I am not sure.

This will be interesting to tackle. I am wondering if the TestHarness is hard-coded to run one test at a time. I don't think that is the case, but now it's got me wondering. @gambka, is this the issue you are concerned with (the additional questions about `parallel_scheduler = true`?


I am heading home for the day, and will address the rest tomorrow!

**Reply by Vandenbg** [reply]
I love the test harness capability. Making it easier to use would be awesome! I know others that use it who would also appreciate this work. 

Here are the contents of that file. To make things a little more complicated, this script calls on a sampler input. Something else you could look at is the fact that the sampler will cue each individual simulation within it using 1 node 1 cpu no matter what. MPI on those would be very helpful... I have to change things up a little bit because I am technically working with KP-Bison, but here it is:
```
[Tests]
  parallel_scheduling = True
  [test1]
    type = 'RunApp'
    input = '../direct_integration.i'
    cli_args = 'input_1d=c1/t_1d.i
                input_cracking=c1/t_cracking.i
                input_asphericity=c1/t_asphericity.i
                output=output/test_out/TEST1'
    max_time = 60
    min_parallel = 48
  []
  [test2]
    type = 'RunApp'
    input = '../direct_integration.i'
    cli_args = 'input_1d=c1/t_1d.i
                input_cracking=c1/t_cracking.i
                input_asphericity=c1/t_asphericity.i
                output=output/test_out/TEST2'
    max_time = 60
    min_parallel = 48
  []
  [test3]
    type = 'RunApp'
    input = '../direct_integration.i'
    cli_args = 'input_1d=c1/t_1d.i
                input_cracking=c1/t_cracking.i
                input_asphericity=c1/t_asphericity.i
                output=output/test_out/TEST3'
    max_time = 60
    min_parallel = 48
  []
[]
```

**Reply by milljm** [reply]
>  Something else you could look at is the fact that the sampler will cue each individual simulation within it using 1 node 1 cpu no matter what. MPI on those would be very helpful... 

I'm afraid I have no idea what you are talking about here. Is this some script in MOOSE somewhere?


I think the only way to handle the wall time is going to be through an additional parameter. A global one:

```pre
[Tests]
  parallel_scheduling = True
  walltime = 5:00:00
  min_parallel = 240
  [test1]
    max_time = 60
    min_parallel = 48
  []
  [test2]
    max_time = 60
    min_parallel = 48
  []
  [test3]
    max_time = 60
    min_parallel = 48
  []
[]
```

If what I want to implement, is implemented, the idea is that 5 jobs would run simultaneously (-j 240 / 48 = 5), and all the jobs contained in this spec file need to finish within 5 hours. Does this sound right?

**Reply by Vandenbg** [reply]
One of these simulations uses the [Quadrature Sampler](https://mooseframework.inl.gov/source/samplers/QuadratureSampler.html). This will run many simulations within it, all with 1 node and 1 cpu. 

Yes that sounds right.

**Reply by milljm** [reply]
Oh, I see. I should let you know that I am not a C++ developer :) So the TestHarness is about it as far as my involvement goes.

**Reply by GiudGiud** [reply]
First step could be implementing those cool things in run_tests logic. Then maybe we would want to factor out the scheduling to a proper "scheduler.py" scripts so people's workflow doesnt involve run_tests
